### ISSUE TITLE
Clarify backwards compatibility policy

### ISSUE THREAD

**Issue by alexkiro** [docker-library/python#964]
As of #954 setuptools was completely removed from the python3.12+ base images, which is not the kinda of change I would have expected from this version.

It would be helpful to have some clarifications on what kinda of changes to expect from the images, and if there are versions of the image that are locked or will be locked so we can update them without risk of breaking everything and still get security updates.

**Comment by alexkiro** [comment]
I image that freezing/locking the current python latest version would be far too limiting for the docker image, and not provide the flexibility to adapt to newer changes. (which I can definitely understand if that was the case) 

My suggestion would be to simply freeze the version before the latest. E.g. right now 3.11 would be locked and when python 3.13 releases later this year images of 3.12.* would be stop receiving any backwards incompatible change, and just continue with patch version updates and security patches/fixes. 

I think there is a significant group of users (myself included) that would rather stay one or two version behind if that gives more stability and lower maintenance. (akin to the practice of having LTS version of software)

**Comment by alexkiro** [comment]
@edmorley Makes a lot of sense; the unpinned setuptools is definetly something great to get rid, and could have potentially cause more damage in the future. If these case are just very rare and low impact, it's probably better to keep the release system simpler rather than complicated them with a freeze/lock timetable.

But if more allowance is needed for the maintainers of this repo, to have the ability to make more breaking changes after python releases, I definetly think that it would be more than reasonable to have a grace period. Maybe something to consider if it ever get to that point :smile: 

Thanks for taking your time to write such a detailed reply!

**Comment by tianon** [comment]
Yeah, we generally try to avoid breaking changes, but in many cases we have to choose between "bad and bad" and oftentimes one ends up being worse for one reason or another, and this was one of those cases (as Ed put _much_ more eloquently than I have :sweat_smile: :heart:).

I know "best effort" isn't _exactly_ what you want to hear, but the fact that it was a surprising event hopefully lends some credence to us generally doing an OK job of avoiding it. :smile:
